### PR TITLE
Resolution and Validation of VC CredentialSchema

### PR DESCRIPTION
**Pull request by nitro-neal**
Resolution and Validation of VC CredentialSchema

This adds the actual fetch and validation of credentialSchema.

kyc schema hosted here - https://purple-charming-snail-690.mypinata.cloud/ipfs/QmZbPpfPXsp4bFQvQvWRaexCu9Vxmj6qkwWerVfPwY9kQS

credential schema validation example:

```
      const vc = await VerifiableCredential.create({
        type: 'KnowYourCustomerCred',
        subject: subjectDid.uri,
        issuer: issuerDid.uri,
        issuanceDate: '2023-05-19T08:02:04Z',
        expirationDate: `2055-05-19T08:02:04Z`,
        data: {
          id: subjectDid.uri,
          country_of_residence: 'US',
          tier: 'Tier 1'
        },
        credentialSchema: {
          id: 'https://purple-charming-snail-690.mypinata.cloud/ipfs/QmZbPpfPXsp4bFQvQvWRaexCu9Vxmj6qkwWerVfPwY9kQS',
          type: 'JsonSchema'
        },
        evidence: [
          { kind: 'document_verification', checks: ['passport', 'utility_bill'] },
          { kind: 'sanctions_check', checks: ['daily'] }
        ]
      });

      const vcJwt = await vc.sign({ did: issuerDid });

      await VerifiableCredential.verify({ vcJwt });
```

This code will now fetch the id in credential schema, and then validate that the credential matches the schema

### CODE REVIEW

**Comment by nitro-neal** [comment]
resolves https://github.com/TBD54566975/web5-js/issues/711

**Review by thehenrytsai — COMMENTED** [review]
Thanks for completing this P1 work item!!

🎖️ 🎸 

Some comments.

File: packages/credentials/src/validators.ts [hunk 1]
```diff
@@ -54,6 +60,35 @@ export class SsiValidator {
     }
   }
 
+  static async validateCredentialSchema(vcDataModel: VcDataModel): Promise<void> {
+    const credentialSchema = vcDataModel.credentialSchema as CredentialSchema;
+
+    if (!credentialSchema || (Array.isArray(credentialSchema) && credentialSchema.length === 0)) {
```
[the remark this answers is not in the document]
Comment (nitro-neal) [reply packages/credentials/src/validators.ts:66]: JS will let me do stuff like this though:

```
      const invalidVcDataModel = { ...validVcDataModel, credentialSchema: [] };
      await SsiValidator.validateCredentialSchema(invalidVcDataModel);
```        

```
         const invalidVcDataModel = { ...validVcDataModel, credentialSchema: undefined };
        await SsiValidator.validateCredentialSchema(invalidVcDataModel);
```
        
        it gets past this line:
        
        ```const credentialSchema = vcDataModel.credentialSchema as CredentialSchema | CredentialSchema[];```

File: packages/credentials/src/validators.ts [hunk 2]
```diff
@@ -54,6 +60,35 @@ export class SsiValidator {
...
+
+      jsonSchema = await response.json();
+    } catch (error: any) {
+      throw new Error(`Failed to fetch schema from ${schemaId}: ${error.message}`);
+    }
+
+    const validationResult = jsonSchemaValidator.validate(vcDataModel, jsonSchema);
+
+    if (!validationResult.valid) {
+      throw new Error(`Schema Validation Errors: ${JSON.stringify(validationResult.errors)}`);
```
[the remark this answers is not in the document]
Comment (nitro-neal) [reply packages/credentials/src/validators.ts:88]: yup thats a good call, added an issue - https://github.com/TBD54566975/web5-js/issues/742

**Comment by frankhinek** [comment]
@nitro-neal for awareness: https://github.com/TBD54566975/schemas/issues/1

**Review by thehenrytsai — APPROVED** [review]
🙏 💯 🚀

**Pull request merged**